tensorflow/python/eager/function_test.py

```diff
@@ -3931,6 +3931,257 @@ def gradients(inputs):
 
     gradients(constant_op.constant([[[1.0], [2.0]]]))  # No error is raised
 
+  def testFollowTypeHintsTraceBasic(self):
+    trace_count = [0]
+    def func(x: ops.Tensor):
+      trace_count[0] += 1
+      return x
+
+    enabled = def_function.function(func, experimental_follow_type_hints=True)
+    disabled = def_function.function(func, experimental_follow_type_hints=False)
+
+    enabled(1) # Initial call gets traced
+    enabled(2)
+    enabled(3)
+    self.assertEqual(trace_count[0], 1)
+
+    trace_count = [0]
+    disabled(1)
+    disabled(2) # Retrace
+    disabled(3) # Retrace
+    self.assertEqual(trace_count[0], 3)
+
+  def testFollowTypeHintsTraceWithArgs(self):
+    trace_count = [0]
+    def func(*args: ops.Tensor):
+      trace_count[0] += 1
+      return args
+
+    enabled = def_function.function(func, experimental_follow_type_hints=True)
+    disabled = def_function.function(func, experimental_follow_type_hints=False)
+
+    args = ("abc", "def",) * 20
+    args2 = ("def", "abc",) * 20
+
+    enabled(args)
+    enabled(args2)
+    self.assertEqual(trace_count[0], 1)
+
+    trace_count = [0]
+    disabled(args)
+    disabled(args2) # Retrace
+    self.assertEqual(trace_count[0], 2)
+
+  def testFollowTypeHintsTraceWithKwargs(self):
+    trace_count = [0]
+    def func(t: ops.Tensor, **kwargs: ops.Tensor):
+      trace_count[0] += 1
+      return t
+
+    enabled = def_function.function(func, experimental_follow_type_hints=True)
+    disabled = def_function.function(func, experimental_follow_type_hints=False)
+
+    enabled(1, x=1, y=1.0, z="one")
+    enabled(2, x=2, y=2.0, z="two")
+    self.assertEqual(trace_count[0], 1)
+
+    trace_count = [0]
+    disabled(1, x=1, y=1.0, z="one")
+    disabled(2, x=2, y=2.0, z="two") # Retrace
+    self.assertEqual(trace_count[0], 2)
+
+  def testFollowTypeHintsTraceWithMultipleInputTypes(self):
+    trace_count = [0]
+    def func(t: ops.Tensor, *args: ops.Tensor,
+             **kwargs: ops.Tensor):
+      trace_count[0] += 1
+      return t
+
+    enabled = def_function.function(func, experimental_follow_type_hints=True)
+    disabled = def_function.function(func, experimental_follow_type_hints=False)
+
+    enabled(1, constant_op.constant(1), "str", x=4.0)
+    enabled(2, constant_op.constant(2), "str2", x=5.0)
+    self.assertEqual(trace_count[0], 1)
+
+    trace_count = [0]
+    disabled(1, constant_op.constant(1), "str", x=4.0)
+    disabled(2, constant_op.constant(2), "str2", x=5.0) # Retrace
+    self.assertEqual(trace_count[0], 2) # pylint: disable=bad-whitespace
```

Comment:
The lint override shouldn't be needed here.

Reply:
There is a `bad-whitespace` pylint error thrown at that line, which should not be happening.

Reply:
It might be for the previous line?  I think pylint wants at least 2 spaces before an inline comment -- i.e., it might want you to add another space before "# Retrace" on line 4009?

Reply:
I added 2 spaces before inline comments but the same pylint error is getting thrown.

Reply:
I think it's caused by these: `i: int = 1,` but the line numbers are wrong. Can you add lint overrides there?